plutus-ir/src/Language/PlutusIR/Compiler/Datatype.hs

```diff
@@ -272,10 +275,12 @@ mkConstructor dty d@(Datatype _ _ tvs _ constrs) index = local (DatatypeComponen
 
     -- case arguments and their types
     casesAndTypes <- do
-          -- these types appear *outside* the scope of the abstraction for the datatype, *but* the name of the datatype will be in
-          -- scope from the wrap, so we can still use it, and we don't need to do anything to the declared types
+          -- these types appear *outside* the scope of the abstraction for the datatype, so we need to use the concrete datatype here
           -- see note [Abstract data types]
+{- MERGE CONFLICT
           let caseTypes = fmap (constructorCaseType (PLC.TyVar p resultType)) constrs
+-}
+          let caseTypes = unveilDatatype (getType dty) d <$> fmap (constructorCaseType (PLC.TyVar p resultType)) constrs
           caseArgNames <- for constrs (\c -> safeFreshName p $ "case_" ++ bsToStr (nameString $ varDeclName c))
```

Comment:
@michaelpj, in your branch it was just `freshName` without `safe` and something else was used in place of `bsToStr`. Doesn't seem like a big deal to me, but please confirm this line shouldn't change.